### PR TITLE
PP-6130 Expunges Criteria Matching Charges

### PR DESCRIPTION
**Pull request by rorymalcolm**
- Edits the test runner cofnig, `test-it-config.yaml` to have expunging
enabled set to true.

- When charges that meet the parity checker criteria between ledger and
connector, and are over a certain number of configuration defined days
old, they will be expunged from the connector database with Ledger
becoming the single point of truth for that charge.

- Adds a number of tests, including modifications to the test runner to
enable wiremock usage with ledger, to ensure this functionality works as
expected.

🤝@alexbishop1

### CODE REVIEW

File: src/main/java/uk/gov/pay/connector/expunge/service/ChargeExpungeService.java [hunk 1]
```diff
@@ -38,14 +46,21 @@ public void expunge(Integer noOfChargesToExpungeQueryParam) {
...
-
-                    logger.info("Charge expunged from connector", kv(PAYMENT_EXTERNAL_ID, chargeEntity.getExternalId()));
-                });
+                    if (hasFinalisedState(chargeEntity)
+                            && parityCheckService.parityCheckChargeForExpunger(chargeEntity)) {
+                        chargeDao.expungeCharge(chargeEntity.getId());
+                        logger.info("Charge expunged from connector {}", 
+                                kv(PAYMENT_EXTERNAL_ID, chargeEntity.getExternalId()));
+                    } else {
+                        chargeDao.updateChargeParityCheckDate(chargeEntity.getId(), ZonedDateTime.now(UTC));
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/connector/expunge/service/ChargeExpungeService.java]: When paritycheck fails charge is already being updated with date and parity check status and also all events are emitted to ledger  https://github.com/alphagov/pay-connector/blob/f6e0a9c1f46ca017552f61ca035ecfceeef53d24/src/main/java/uk/gov/pay/connector/tasks/ParityCheckService.java#L89. Either ParityCheckService can be updated or logic extracted here to update charge & emit events

We need to find a way to distinguish logs that the charge has been tried for expunging before and can't be expunge on retry. This can be then be used for alerting to fix ParityChecker / Ledger manually or investigate on these charges.

File: src/test/java/uk/gov/pay/connector/expunge/service/ChargeExpungeServiceTest.java [hunk 2]
```diff
@@ -66,4 +73,24 @@ public void expunge_shouldNotExpungeChargesIfFeatureIsNotEnabled() {
...
+        chargeEntity.setStatus(ChargeStatus.AUTHORISATION_READY);
+        chargeEntity.setStatus(ChargeStatus.AUTHORISATION_SUCCESS);
+        chargeEntity.setStatus(ChargeStatus.CAPTURE_READY);
+        chargeEntity.setStatus(ChargeStatus.CAPTURE_SUBMITTED);
+        when(mockExpungeConfig.isExpungeChargesEnabled()).thenReturn(true);
+        when(mockExpungeConfig.getMinimumAgeOfChargeInDays()).thenReturn(minimumAgeOfChargeInDays);
+        when(mockChargeDao.findChargeToExpunge(minimumAgeOfChargeInDays, defaultExcludeChargesParityCheckedWithInDays))
+                .thenReturn(Optional.of(chargeEntity));
+        when(parityCheckService.parityCheckChargeForExpunger(chargeEntity)).thenReturn(true);
+        chargeExpungeService.expunge(2);
```
Comment (kbottla) [inline src/test/java/uk/gov/pay/connector/expunge/service/ChargeExpungeServiceTest.java:85]: missing some assertions (`verify()`) here ?

File: src/test/java/uk/gov/pay/connector/expunge/service/ChargeExpungeServiceTest.java [hunk 3]
```diff
@@ -66,4 +73,24 @@ public void expunge_shouldNotExpungeChargesIfFeatureIsNotEnabled() {
...
+        ChargeEntity chargeEntity = aWebChargeEntity()
+                .withGatewayAccount(aGatewayAccountEntity().withId(1L).build())
+                .withAmount(120L)
+                .build();
+        chargeEntity.setStatus(ChargeStatus.CREATED);
+        chargeEntity.setStatus(ChargeStatus.ENTERING_CARD_DETAILS);
+        chargeEntity.setStatus(ChargeStatus.AUTHORISATION_READY);
+        chargeEntity.setStatus(ChargeStatus.AUTHORISATION_SUCCESS);
+        chargeEntity.setStatus(ChargeStatus.CAPTURE_READY);
+        chargeEntity.setStatus(ChargeStatus.CAPTURE_SUBMITTED);
```
Comment (kbottla) [inline src/test/java/uk/gov/pay/connector/expunge/service/ChargeExpungeServiceTest.java]: Any reason setting multiple statuses are set for a charge ?

File: src/test/java/uk/gov/pay/connector/it/resources/ExpungeResourceIT.java [hunk 4]
```diff
@@ -1,37 +1,237 @@
...
+        this.defaultTestAccount = DatabaseFixtures
+                .withDatabaseTestHelper(databaseTestHelper)
+                .aTestAccount()
+                .withAccountId(nextLong())
+                .insert();
+    }
 
+
+    @Test
+    public void shouldExpireChargesBeforeAndAfterAuthorisationAndShouldHaveTheRightEvents() throws JsonProcessingException {
```
Comment (kbottla) [inline src/test/java/uk/gov/pay/connector/it/resources/ExpungeResourceIT.java]: Not a mistake in this PR but name should be changed to match `epunge....`

File: src/test/java/uk/gov/pay/connector/it/resources/ExpungeResourceIT.java [hunk 5]
```diff
@@ -1,37 +1,237 @@
...
+                .statusCode(200);
+        var postCharge = databaseTestHelper.containsChargeWithExternalId("external_charge_id_2");
+        assertThat(postCharge, is(true));
     }
 
+    @Test
+    public void shouldExpungeSomeChargesButNotOthers() throws JsonProcessingException {
+        ledgerStub = new LedgerStub();
+
+        testCharge = DatabaseFixtures.withDatabaseTestHelper(databaseTestHelper)
```
Comment (kbottla) [inline src/test/java/uk/gov/pay/connector/it/resources/ExpungeResourceIT.java]: can this be named such that they indicate why they won't be expunged . It is a hard to understand in this test why `some charges` are expunges vs the rest

File: src/main/java/uk/gov/pay/connector/charge/dao/ChargeDao.java [hunk 6]
```diff
@@ -211,4 +211,18 @@ public Long findMaxId() {
                 .setMaxResults(1)
                 .getResultList().stream().findFirst();
     }
+
+    public void expungeCharge(Long id) {
+        ChargeEntity chargeEntity = entityManager.get().find(ChargeEntity.class, id);
+        entityManager.get().remove(chargeEntity);
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/connector/charge/dao/ChargeDao.java:217]: Sorry missed this in first round. We need to explicitly remove from emitted_events, tokens (although we delete tokens older than x days but good to be explicit)  as there is no foreign key relation.
_I am guessing entityManager.remove() cascades delete operation ?_

Happy to be addressed in separate PR though

File: src/main/java/uk/gov/pay/connector/charge/model/domain/ChargeStatus.java [hunk 7]
```diff
@@ -16,59 +16,65 @@
...
+    AWAITING_CAPTURE_REQUEST("AWAITING CAPTURE REQUEST", EXTERNAL_CAPTURABLE, false),
+    CAPTURE_APPROVED("CAPTURE APPROVED", EXTERNAL_SUCCESS, false),
+    CAPTURE_APPROVED_RETRY("CAPTURE APPROVED RETRY", EXTERNAL_SUCCESS, false),
+    CAPTURE_READY("CAPTURE READY", EXTERNAL_SUCCESS, false),
+    CAPTURED("CAPTURED", EXTERNAL_SUCCESS, true),
+    CAPTURE_SUBMITTED("CAPTURE SUBMITTED", EXTERNAL_SUCCESS, false),
+    CAPTURE_ERROR("CAPTURE ERROR", EXTERNAL_ERROR_GATEWAY, true),
+
+    EXPIRE_CANCEL_READY("EXPIRE CANCEL READY", EXTERNAL_FAILED_EXPIRED, false),
+    EXPIRE_CANCEL_FAILED("EXPIRE CANCEL FAILED", EXTERNAL_FAILED_EXPIRED, false),
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/connector/charge/model/domain/ChargeStatus.java]: Charges with state `EXPIRE_CANCEL_FAILED ` can be expunged

File: src/main/java/uk/gov/pay/connector/charge/model/domain/ChargeStatus.java [hunk 8]
```diff
@@ -16,59 +16,65 @@
...
+    EXPIRE_CANCEL_SUBMITTED("EXPIRE CANCEL SUBMITTED", EXTERNAL_FAILED_EXPIRED, false),
+    EXPIRED("EXPIRED", EXTERNAL_FAILED_EXPIRED, true),
+
+    SYSTEM_CANCEL_READY("SYSTEM CANCEL READY", EXTERNAL_CANCELLED, false),
+    SYSTEM_CANCEL_ERROR("SYSTEM CANCEL ERROR", EXTERNAL_CANCELLED, false),
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/connector/charge/model/domain/ChargeStatus.java]: same here

File: src/main/java/uk/gov/pay/connector/charge/model/domain/ChargeStatus.java [hunk 9]
```diff
@@ -16,59 +16,65 @@
...
+    SYSTEM_CANCEL_SUBMITTED("SYSTEM CANCEL SUBMITTED", EXTERNAL_CANCELLED, false),
+    SYSTEM_CANCELLED("SYSTEM CANCELLED", EXTERNAL_CANCELLED, true),
+
+    USER_CANCEL_READY("USER CANCEL READY", EXTERNAL_FAILED_CANCELLED, false),
+    USER_CANCEL_SUBMITTED("USER CANCEL SUBMITTED", EXTERNAL_FAILED_CANCELLED, false),
+    USER_CANCELLED("USER CANCELLED", EXTERNAL_FAILED_CANCELLED, false),
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/connector/charge/model/domain/ChargeStatus.java]: this too

File: src/main/java/uk/gov/pay/connector/expunge/service/ChargeExpungeService.java [hunk 10]
```diff
@@ -59,4 +54,30 @@ private int getNumberOfChargesToExpunge(Integer noOfChargesToExpungeQueryParam)
...
+    private void parityCheckAndExpungeIfMet(ChargeEntity chargeEntity) {
+        if (!inTerminalState(chargeEntity)) {
+            logger.info("Charge not expunged because it is not in a terminal state {}",
+                    kv(PAYMENT_EXTERNAL_ID, chargeEntity.getExternalId()));
+        } else if (parityCheckService.parityCheckChargeForExpunger(chargeEntity)) {
+            chargeDao.expungeCharge(chargeEntity.getId());
+            logger.info("Charge expunged from connector {}",
+                    kv(PAYMENT_EXTERNAL_ID, chargeEntity.getExternalId()));
+        } else {
+            if (chargeEntity.getParityCheckDate() == null) {
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/connector/expunge/service/ChargeExpungeService.java]: `parityCheckDate` is never null here as charge would have been updated already https://github.com/alphagov/pay-connector/blob/f6e0a9c1f46ca017552f61ca035ecfceeef53d24/src/main/java/uk/gov/pay/connector/tasks/ParityCheckService.java#L99

Also mentioned in https://github.com/alphagov/pay-connector/pull/2094#discussion_r377689868

File: src/main/java/uk/gov/pay/connector/expunge/service/ChargeExpungeService.java [hunk 11]
```diff
@@ -6,10 +6,15 @@
 import uk.gov.pay.connector.app.config.ExpungeConfig;
 import uk.gov.pay.connector.charge.dao.ChargeDao;
 import uk.gov.pay.connector.charge.model.domain.ChargeEntity;
+import uk.gov.pay.connector.charge.model.domain.ChargeStatus;
+import uk.gov.pay.connector.tasks.ParityCheckService;
 
 import javax.inject.Inject;
+import java.time.ZonedDateTime;
 import java.util.Optional;
+import java.util.stream.IntStream;
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/connector/expunge/service/ChargeExpungeService.java:14]: Unused import `java.util.Optional;`

File: src/test/java/uk/gov/pay/connector/it/resources/ExpungeResourceIT.java [hunk 12]
```diff
@@ -1,37 +1,237 @@
...
+        this.defaultTestAccount = DatabaseFixtures
+                .withDatabaseTestHelper(databaseTestHelper)
+                .aTestAccount()
+                .withAccountId(nextLong())
+                .insert();
+    }
 
+
+    @Test
+    public void shouldExpungeChargesBeforeAndAfterAuthorisationAndShouldHaveTheRightEvents() throws JsonProcessingException {
```
Comment (kbottla) [inline src/test/java/uk/gov/pay/connector/it/resources/ExpungeResourceIT.java]: rename ? `shouldExpungeCharge`~~`sBeforeAndAfterAuthorisationAndShouldHaveTheRightEvents`~~

File: src/main/java/uk/gov/pay/connector/charge/dao/ChargeDao.java [hunk 13]
```diff
@@ -211,4 +211,18 @@ public Long findMaxId() {
...
+    }
+
+    public void updateChargeParityCheckDate(Long id, ZonedDateTime dateTime) {
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/connector/charge/dao/ChargeDao.java]: This may not be needed depending on approach taken for https://github.com/alphagov/pay-connector/pull/2094/files#r378099308

File: src/main/java/uk/gov/pay/connector/expunge/service/ChargeExpungeService.java (hunk 10 above)
Comment (rorymalcolm) [reply src/main/java/uk/gov/pay/connector/expunge/service/ChargeExpungeService.java]: We've addressed this in the PR

File: src/main/java/uk/gov/pay/connector/charge/dao/ChargeDao.java (hunk 6 above)
Comment (rorymalcolm) [reply src/main/java/uk/gov/pay/connector/charge/dao/ChargeDao.java:217]: gonna do it in a separate PR

Comment (sfount) [reply src/main/java/uk/gov/pay/connector/charge/dao/ChargeDao.java:217]: +1 making sure charge events, emitted events and tokens are removed here -- opening this thread for visibility so we don't forget. Can be closed by the next PR.

**Review by kbottla — APPROVED** [review]
